src/main/java/org/terasology/moduletestingenvironment/TestEventReceiver.java

```diff
@@ -63,64 +74,79 @@
+ * <p>
+ * Note that listeners are discarded with the rest of the engine between tests, so closing your receiver is only useful
+ * if you need to stop handling events within a single test method.
  */
-public class TestEventReceiver<T extends Event> implements AutoCloseable, EventReceiver<T>{
+public class TestEventReceiver<T extends Event> implements AutoCloseable, EventReceiver<T> {
+    private final EventSystem eventSystem;
+    private final Class<T> eventClass;
+    private final BiConsumer<T, EntityRef> handler;
 
-    private EventSystem eventSystem;
-    private Class<T> eventClass;
-    private List<EntityRef> entityRefs = new ArrayList<>();
-    private List<T> events = new ArrayList<>();
-    private BiConsumer<T, EntityRef> handler = (a, b)-> { };
+    private final List<EntityRef> entityRefs = new ArrayList<>();
+    private final List<T> events = new ArrayList<>();
 
     /**
      * Constructs a new {@code TestEventReceiver} and registers it to listen for events.
-     *
+     * <p>
+     * The following signature of a {@code TestEventReceiver} is equivalent to the event handler below:
      * <pre>
-     * {@literal
-     * TestEventReceiver receiver = new TestEventReceiver<>(context, DropItemEvent.class, (event, entity) -> {
+     * {@code
+     * TestEventReceiver receiver = new TestEventReceiver<>(context, MyEvent.class, (event, entity) -> {
      *   // do something with the event or entity
-     * });
+     * }, MyComponent.class);
+     *
+     * // ... corresponds to
+     *
+     * @ReceiveEvent(components = {MyComponent.class})
+     * public void handler(MyEvent event, EntityRef entity) {
+     *     // do something with the event or entity
+     * }
      * }
      * </pre>
      *
-     * @param context the context object for the test; this should probably be obtained through
-     *                {@link ModuleTestingEnvironment#getHostContext()} and is needed so we can
-     *                obtain an {@link EventSystem} instance to register our event handler.
-     * @param eventClass   the {@link Event} subclass to listen for
-     * @param handler an optional {@link BiConsumer} to fire when events are received.
+     * @param context the context object for the test; this should probably be obtained through {@link
+     *         ModuleTestingHelper#createClient()} and is needed so we can obtain an {@link EventSystem} instance to
+     *         register our event handler.
+     * @param eventClass the {@link Event} subclass to listen for
+     * @param handler an optional {@link BiConsumer} fired when events are received
+     * @param componentTypes list of component types that need to be present on the entity receiving the event
      */
-    public TestEventReceiver(Context context, Class<T> eventClass, BiConsumer<T, EntityRef> handler) {
-        this.handler = handler;
+    public TestEventReceiver(Context context, Class<T> eventClass, BiConsumer<T, EntityRef> handler, Class<?
```

Comment:
This line now has a warning about the use of varargs:

> Possible heap pollution from parameterized vararg type 

I'm not really sure whether we want to use varargs here or not, but inherently the underlying `EventSystem` only exposes API with varargs 🤔  I can only assume this is about "type-safe" arrays without the overhead of having a `List<T>`. 
However, we may want to offer list-based variants of `EventSystem::registerEventReceiver` in addition to the varargs ones. Casting/transforming between the different collection-likes is still awkward, though...